// react-spatial-navigation/lib/decorators/Element.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _spatialNavigation = require('spatial-navigation');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var withElement = function withElement(settings) {
  return function (ComposedComponent) {
    var SNElementComponent = function (_React$Component) {
      _inherits(SNElementComponent, _React$Component);

      function SNElementComponent() {
        _classCallCheck(this, SNElementComponent);

        var _this = _possibleConstructorReturn(this, (SNElementComponent.__proto__ || Object.getPrototypeOf(SNElementComponent)).call(this));

        _this.handleChildRef = function (component) {
          if (!component) {
            return;
          }

          var state = component.state,
              props = component.props;


          var ctxParent = _this.context.element || _this.context.container;
          var domEl = _reactDom2.default.findDOMNode(component);

          _this.element.connectDomEl(domEl);

          var disabled = state && state.disabled || props && props.disabled;

          if (disabled) {
            _this.element.disable();
          }

          if (Boolean(ctxParent.collection)) {
            ctxParent.collection.add(_this.element);
          }
        };

        _this.element = _spatialNavigation.Element.create(null, settings);

        _this.state = {
          SNElement: _this.element
        };
        return _this;
      }

      _createClass(SNElementComponent, [{
        key: 'getChildContext',
        value: function getChildContext() {
          return {
            element: this.element
          };
        }
      }, {
        key: 'render',
        value: function render() {
          return _react2.default.createElement(ComposedComponent, _extends({
            ref: this.handleChildRef
          }, this.props, this.state));
        }
      }]);

      return SNElementComponent;
    }(_react2.default.Component);

    SNElementComponent.displayName = "SN:Element";
    SNElementComponent.contextTypes = {
      container: _react2.default.PropTypes.instanceOf(_spatialNavigation.Container),
      element: _react2.default.PropTypes.instanceOf(_spatialNavigation.Element)
    };


    SNElementComponent.childContextTypes = {
      element: _react2.default.PropTypes.instanceOf(_spatialNavigation.Element)
    };

    return SNElementComponent;
  };
};

exports.default = withElement;
