// spatial-navigation/lib/EventAggregator.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EventAggregator = exports.EVENT_PREFIX = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _Event = require('./Event');

var _Event2 = _interopRequireDefault(_Event);

var _constants = require('./util/constants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var EventAggregator = function () {
  function EventAggregator() {
    _classCallCheck(this, EventAggregator);

    this.events = [];
  }

  _createClass(EventAggregator, [{
    key: 'getEvent',
    value: function getEvent(eventName) {
      return this.events.find(function (_ref) {
        var name = _ref.name;
        return name === eventName;
      });
    }
  }, {
    key: 'dispatchEvent',
    value: function dispatchEvent(eventName) {
      var _event;

      var event = this.getEvent(eventName);

      if (!event) {
        event = new _Event2.default(eventName);
        this.events.push(event);
      }

      for (var _len = arguments.length, eventArgs = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
        eventArgs[_key - 1] = arguments[_key];
      }

      (_event = event).call.apply(_event, eventArgs);
    }

    /**
     *
     * @param eventName {string}
     * @param handler {function}
     * @param priority {string|number}
     * @returns {{ event: *, handler: * }}
     */

  }, {
    key: 'subscribe',
    value: function subscribe(eventName, handler, priority) {
      var event = this.getEvent(eventName);

      if (!event) {
        event = new _Event2.default(eventName);
        this.events.push(event);
      }

      event.addHandler(handler, priority);

      return { event: event, handler: handler };
    }

    /**
     *
     * @param eventName {string}
     * @param handler {function}
     * @param priority {string|number}
     * @returns {{ event: *, handlerWrapper: (function(...[*])) }}
     */

  }, {
    key: 'once',
    value: function once(eventName, handler, priority) {
      var event = this.getEvent(eventName);

      if (!event) {
        event = new _Event2.default(eventName);
        this.events.push(event);
      }

      var handlerWrapper = function handlerWrapper() {
        var result = handler.apply(undefined, arguments);
        if (result) {
          event.removeHandler(handlerWrapper, priority);
        }
      };

      event.addHandler(handlerWrapper, priority);

      return { event: event, handlerWrapper: handlerWrapper };
    }
  }]);

  return EventAggregator;
}();

exports.default = new EventAggregator();
exports.EVENT_PREFIX = _constants.EVENT_PREFIX;
exports.EventAggregator = EventAggregator;
