// spatial-navigation/lib/KeyMapNavigation.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _EventAggregator = require('./EventAggregator');

var _EventAggregator2 = _interopRequireDefault(_EventAggregator);

var _Navigation = require('./Navigation');

var _Navigation2 = _interopRequireDefault(_Navigation);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var KeyMapNavigation = function () {
  function KeyMapNavigation() {
    _classCallCheck(this, KeyMapNavigation);

    this.map = {};

    this.bindListeners();
  }

  _createClass(KeyMapNavigation, [{
    key: 'bindListeners',
    value: function bindListeners() {
      _EventAggregator2.default.subscribe(_EventAggregator.EVENT_PREFIX + 'keypress', this.navigate.bind(this));
    }
  }, {
    key: 'navigate',
    value: function navigate(keyCombination) {
      var instanceToNavigate = this.getInstance(keyCombination);

      if (instanceToNavigate) {
        var prevFocusedElement = _Navigation2.default.focusedElement;

        var _EA$subscribe = _EventAggregator2.default.subscribe(_EventAggregator.EVENT_PREFIX + 'esc', function () {
          prevFocusedElement.focus();
        }),
            event = _EA$subscribe.event,
            handler = _EA$subscribe.handler;

        _EventAggregator2.default.once(_EventAggregator.EVENT_PREFIX + 'blurElement', function (element) {
          if (instanceToNavigate == element) {
            event.removeHandler(handler);
            return true;
          }
        });

        instanceToNavigate.focus();
      }
    }
  }, {
    key: 'addRelation',
    value: function addRelation(map, instance) {
      for (var keyCombination in map) {
        this.map[keyCombination] = instance;
      }
    }
  }, {
    key: 'getInstance',
    value: function getInstance(keyCombination) {
      return this.map[keyCombination];
    }
  }]);

  return KeyMapNavigation;
}();

exports.default = new KeyMapNavigation();
module.exports = exports['default'];
