// spatial-navigation/lib/Logger.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LEVEL_ERROR = exports.LEVEL_WARN = exports.LEVEL_INFO = exports.LEVEL_DEBUG = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _constants = require('./util/constants');

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var LEVEL_DEBUG = exports.LEVEL_DEBUG = 0;
var LEVEL_INFO = exports.LEVEL_INFO = 1;
var LEVEL_WARN = exports.LEVEL_WARN = 2;
var LEVEL_ERROR = exports.LEVEL_ERROR = 3;

var Logger = function () {
  function Logger() {
    _classCallCheck(this, Logger);

    this.debug = Logger.log.bind(this, LEVEL_DEBUG);
    this.info = Logger.log.bind(this, LEVEL_INFO);
    this.warn = Logger.log.bind(this, LEVEL_WARN);
    this.error = Logger.log.bind(this, LEVEL_ERROR);
  }

  _createClass(Logger, null, [{
    key: 'log',
    value: function log(level, msg, attributes) {
      if (Logger.getLevel(_constants.LOG_LEVEL) > level) return;
      console.log(Logger.getLevelView(level) + ' - ' + msg, attributes);
    }
  }, {
    key: 'getLevel',
    value: function getLevel(strLevel) {
      return ['debug', 'info', 'warn', 'error'].find(function (level) {
        return level == strLevel.toLowerCase();
      });
    }

    /**
     *
     * @param level
     * @returns {string}
     */

  }, {
    key: 'getLevelView',
    value: function getLevelView(level) {
      return ['debug', 'info', 'warn', 'error'][level];
    }
  }]);

  return Logger;
}();

exports.default = new Logger();
